proof/invariant-abstract/IpcDet_AI.thy

```diff
@@ -1086,6 +1086,180 @@ lemma reply_push_sender_sc_Some_invs:
+    and K (tcb_ptr = tcb_ptr' \<longrightarrow>
+           (\<forall>tcb x. proj (tcb_to_itcb (tcb_sched_context_update x tcb)) = proj (tcb_to_itcb tcb)))\<rbrace>
+   maybe_return_sc ntfn_ptr tcb_ptr
+   \<lbrace>\<lambda>rv s. Q (pred_tcb_at proj P tcb_ptr' s)\<rbrace>"
+  apply (wpsimp simp: maybe_return_sc_def set_tcb_obj_ref_def set_object_def
+                      get_tcb_obj_ref_def thread_get_def get_sk_obj_ref_def get_simple_ko_def
+                      get_object_def)
+  apply (auto simp: pred_tcb_at_def obj_at_def get_tcb_SomeD)
+  done
+
+lemma maybe_return_sc_st_tcb_at[wp]:
+  "maybe_return_sc ntfn_ptr tcb_ptr \<lbrace>st_tcb_at P t\<rbrace>"
+  by (wpsimp wp: maybe_return_sc_pred_tcb_at)
+
+(* FIXME RT: move *)
+lemma receive_ipc_preamble_rv_lift:
+  "(\<And>a P. f \<lbrace>reply_tcb_reply_at P a \<rbrace>)
+   \<Longrightarrow> (\<And>a P. f \<lbrace>reply_sc_reply_at P a \<rbrace>)
+   \<Longrightarrow> f \<lbrace>receive_ipc_preamble_rv reply reply_opt\<rbrace>"
+  by (case_tac reply_opt; wpsimp wp: hoare_vcg_conj_lift; assumption)
+
+lemma maybe_return_sc_fault_tcbs_valid_states[wp]:
+  "maybe_return_sc y thread \<lbrace>fault_tcbs_valid_states\<rbrace>"
+  unfolding fault_tcbs_valid_states_def
+  by (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift' maybe_return_sc_pred_tcb_at)
+
+crunches maybe_return_sc
+  for misc_proj[wp]: "\<lambda>s. P (cur_thread s) (cur_sc s) (release_queue s)
+                                 (cur_domain s) (domain_time s) (idle_thread s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+crunches reschedule_required
+  for sc_at_ppred[wp]: "sc_at_ppred proj P scp"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma maybe_return_sc_sc_at_ppred:
+  "\<lbrace>sc_at_ppred proj P scp and bound_sc_tcb_at(\<lambda>x. x \<noteq> Some scp) tcb_ptr\<rbrace>
+   maybe_return_sc ntfn_ptr tcb_ptr
+   \<lbrace>\<lambda>rv. sc_at_ppred proj P scp\<rbrace>"
+  apply (wpsimp simp: maybe_return_sc_def update_sched_context_def
+                      set_tcb_obj_ref_def set_object_def
+                      get_tcb_obj_ref_def thread_get_def get_sk_obj_ref_def get_simple_ko_def
+                      get_object_def)
+  by (auto simp: sc_at_pred_n_def obj_at_def pred_tcb_at_def get_tcb_SomeD)
+
+lemma reschedule_required_cur_sc_tcb':
+  "\<lbrace>\<lambda>s. sc_tcb_sc_at \<top> (cur_sc s) s\<rbrace> reschedule_required \<lbrace>\<lambda>_. cur_sc_tcb\<rbrace>"
+  supply set_scheduler_action_cur_sc_tcb [wp del]
+  unfolding reschedule_required_def
+  by (wpsimp simp: reschedule_required_def set_scheduler_action_def tcb_sched_action_def
+                   set_tcb_queue_def get_tcb_queue_def thread_get_def is_schedulable_def
+                   cur_sc_tcb_def sc_tcb_sc_at_def obj_at_def)
+
+lemma maybe_return_sc_cur_sc_tcb[wp]:
+  "\<lbrace>cur_sc_tcb and (\<lambda>s. thread = cur_thread s)\<rbrace>
+    maybe_return_sc y thread
+   \<lbrace>\<lambda>_. cur_sc_tcb\<rbrace>"
```

Comment:
I guess this is where the new `thread = cur_thread s` precondition comes from in `ri_invs`.

I think it might be possible to remove this precondition, and replace it with `sym_refs`:
- If the `scheduler_action` is *not* `resume_cur_thread`, then the postcondition is satisfied as long as there is a scheduling context at `cur_sc`, which you get from the precondition. No problem here.
- If the `scheduler_action` *is* `resume_cur_thread`, and `thread` is `cur_thread`, then the `reschedul_required` means that the postcondition is similarly satisfied. Still no problem.
- But if the `scheduler_action` is `resume_cur_thread`, and `thread` is *not* `cur_thread`, then you would need some extra precondition to establish that the scheduling context that's updated is not the `cur_sc`. I think `sym_refs` would give you that, since you know that the `cur_sc` is bound to `cur_thread`, and `thread` is bound to the scheduling context that's updated, but `thread` is not `cur_thread`.

I can see why you did this, though: `maybe_return_sc` is only called from `receive_signal` and now `receive_ipc`, which are only called on the `cur_thread`. And that's easier to reason about than `sym_refs`.

Even so, it might be worth thinking about whether it's worth replacing the precondition. If not, consider whether `[wp]` is appropriate for such a strong precondition.

Finally, if we're committing to only calling `maybe_return_sc` on the `cur_thread`, why are we testing for this in the implementation? (As opposed to asserting it.)

Reply:
Also, I think this is obscure enough to warrant a comment on why this precondition is needed, especially if you don't change the precondition to `sym_refs`.

Also also, meaningful variable names (here and in some other lemmas).

Reply:
I like that `sym_refs` seems strong enough but it might be harder to use in proofs that use this wp rule.

I think the best option here is to remove `[wp]`, keep the existing precondition, and add some comments to explain. Sound ok?

Reply:
I was going to say you could use `heap_refs_retract_at`, but that won't be available in `IpcDet_AI`. :-(

I'll leave the decision to you, but I'm pretty sure you have all of `invs` at the point you need this, so I don't think `sym_refs` should cause any trouble.

I still think it would also be good to look into why the `reschedule_required` in `maybe_return_sc` is conditional on whether `thread` is the current thread. This lemma raises the question of whether we ever call it on anything other than the current thread. From what I can tell, we don't. In that case, the `thread = cur_thread s` could be an assertion instead of a precondition, and the `reschedule_required` could be unconditional. Refine will force us to (easily) prove the assertion, and we'll get the assurance that it's only called in the way we expect. I'll make a separate comment on that bit of the patch.

Reply:
Oh, and don't forget to give things names that are somewhat more meaningful than `y`! :-)

Reply:
Thanks! You're right on both counts: the check is not needed at all and replacing the precondition with sym_refs seems to work just fine.